// serverless-lex-model/index.js
'use strict';

var AWS = require('aws-sdk');
AWS.config.setPromisesDependency(require('bluebird'));
var lexModelHelper = require('./lex-model-helper');

class ServerlessLexModel {
  constructor(serverless, options) {
    this.serverless = serverless;
    this.options = options;

    this.commands = {
      lex: {
        lifecycleEvents: [
          'deploy',
        ]
      },
    };

    this.hooks = {
      'lex:deploy': this.afterDeploy.bind(this),
      'before:deploy:initialize': this.afterDeploy.bind(this),
      // 'after:deploy:deploy': this.afterDeploy.bind(this),
    }
  }

  afterDeploy() {
    this.serverless.cli.log('Deploying a Lex bot...');
    lexModelHelper.setRegion(this.serverless.service.provider.region);
    var lexConfig = this.serverless.service.custom.lex;
    this.getBot(lexConfig.name);
  }

  getBot(name) {
    lexModelHelper.getBot(name).then(function(data) {
      this.putBot(name, data.checksum);
    }.bind(this)).catch(function(error) {
      if (error.code === 'NotFoundException') {
        this.putBot(name, null);
      } else {
        this.serverless.cli.log(error.message);
      }
    }.bind(this));
    // lexModelHelper.putBot(lexConfig.name, null);
  }

  putBot(name, checksum) {
    lexModelHelper.putBot(name, checksum).then(function(data) {
      this.serverless.cli.log(`${data.name} bot has been successfully created!`);
    }.bind(this)).catch(function(error) {
      this.serverless.cli.log(error.message);
    }.bind(this));
  }
}

module.exports = ServerlessLexModel;
