// serverless-lex-model/lex-model-helper.js
var AWS = require('aws-sdk');
var region = AWS.config.update({region:'us-east-1'});
var lex = new AWS.LexModelBuildingService();
// lexmodelbuildingservice.createBotVersion(params, function (err, data) {
//   if (err) console.log(err, err.stack); // an error occurred
//   else     console.log(data);           // successful response
// });

setRegion = function(region) {
  region = region
}


getBot = function (name) {
  var params = {
   name: name,
   versionOrAlias: "$LATEST"
  };
  return lex.getBot(params).promise();
};

putBot = function (name, checksum) {
  var params = {
    name: name,
    description: "Lex for the MMT Slack bot.",
    intents: [
       {
      intentName: "TodaysWeather",
      intentVersion: "$LATEST"
     }
    ],
    clarificationPrompt: {
     maxAttempts: 1,
     messages: [
        {
       content: "I'm sorry, I didn't hear that. Can you repeate what you just said?",
       contentType: "PlainText"
      },
        {
       content: "Can you say that again?",
       contentType: "PlainText"
      }
     ]
    },
    abortStatement: {
     messages: [
        {
       content: "I don't understand. Can you try again?",
       contentType: "PlainText"
      },
        {
       content: "I'm sorry, I don't understand.",
       contentType: "PlainText"
      }
     ]
    },
    idleSessionTTLInSeconds: 300,
    processBehavior: "BUILD",
    locale: "en-US",
    childDirected: false
  };
  // Add checksum params if the bot is arleay exist.
  if (checksum !== null) {
    params['checksum'] = checksum
  }
  return lex.putBot(params).promise();
};

module.exports = {
  setRegion,
  getBot,
  putBot
};
